write-fonts/src/tables/cmap.rs

```diff
@@ -2,62 +2,142 @@
 //!
 //! [cmap]: https://docs.microsoft.com/en-us/typography/opentype/spec/cmap
 
+use std::collections::HashSet;
+
 include!("../../generated/generated_cmap.rs");
 
+// https://learn.microsoft.com/en-us/typography/opentype/spec/cmap#unicode-platform-platform-id--0
+const UNICODE_BMP_ENCODING: u16 = 3;
+
+fn size_of_cmap4(seg_count: u16, gid_count: u16) -> u16 {
+    8 * 2  // 8 uint16's
+    + 2 * seg_count * 4  // 4 parallel arrays of len seg_count, 2 bytes per entry
+    + 2 * gid_count // 2 bytes per gid in glyphIdArray
+}
+
+impl CmapSubtable {
+    fn create_format_4(
+        lang: u16,
+        end_code: Vec<u16>,
+        start_code: Vec<u16>,
+        id_deltas: Vec<i16>,
+    ) -> Self {
+        if HashSet::from([start_code.len(), end_code.len(), id_deltas.len()]).len() != 1 {
+            panic!("uneven parallel arrays, very bad. Very very bad.");
+        }
+
+        let seg_count: u16 = start_code.len().try_into().unwrap();
+        // Spec: Log2 of the maximum power of 2 less than or equal to segCount (log2(searchRange/2),
+        // which is equal to floor(log2(segCount)))
+        let entry_selector = (seg_count as f32).log2().floor();
+
+        // Spec: Maximum power of 2 less than or equal to segCount, times 2
+        // ((2**floor(log2(segCount))) * 2, where “**” is an exponentiation operator)
+        let search_range = 2u16.pow(entry_selector as u32).checked_mul(2).unwrap();
+
+        // if 2^entry_selector*2 is a u16 then so is entry_selector
+        let entry_selector: u16 = entry_selector as u16;
+        let range_shift: u16 = seg_count * 2 - search_range;
```

Comment:
```suggestion
        let entry_selector = entry_selector as u16;
        let range_shift = seg_count * 2 - search_range;
```

The type here is trivially inferred, so I don't see much advantage in declaring it explicitly?